### ISSUE TITLE
Recursion via helper function that preserves/reduces size is still rejected by termination checker

### ISSUE THREAD

**Issue by fbanados** [ionathanch/coq#13]
#### Description of the problem

```coq
Unset Guard Checking.
Set Sized Typing.
Definition tl {A : Type} (l : list A) : list A :=
  match l with
  | cons hd tl => tl
  | nil => nil
  end.
Fail Fixpoint filter {A : Type} (f : A -> bool)
         (l : list A) {struct l}: list A :=
  match l with
  | nil => nil
  | cons x l0 => if f x then cons x (filter f (tl l)) else filter f (tl l)
  end.
```
If you don't annotate the fixpoint with (`{struct l}`), fails with `Cannot guess decreasing argument of fix` standard error. 
With the annotation, fails with the following message:
```
Error: The size annotation of the recursive type in the
(co)fixpoint
"fix filter (A : Type) (f : A -> bool) (l : list A) {struct l} : list A := match l with
                                                                           | nil => nil
                                                                           | (x :: _)%list => if f x then (x :: filter A f (tl l))%list else filter A f (tl l)
                                                                           end"
could not be set to a finite size variable.
```

Shouldn't this definition be accepted?

#### Coq Version

eec3706d80615473046b67ebe565b608d35a6e60

**Comment by fbanados** [comment]
Is there a timeline estimate for this bug?

**Comment by ionathanch** [comment]
I've taken a closer look at this and unfortunately our system is *not* sophisticated enough to accept this definition, unfortunately. The best we can do for `tl` is for it to be size preserving, i.e. `tl: List^{v+1} A -> List^{v+1} A`, for instance. Supposing that the parameter `l` of `filter` has type `List^{w+1}`, we have `(tl l): List^{w+1}`, but in `(filter f (tl l))`, we need the recursive call to be on a smaller argument of size at most `List^w`.
<hr>

Technical notes:
The solver solves for the *smallest* solution for all size variables, which results in `tl: (A: Type) -> (l: List^{s8} A) -> List^{s8+1} A`. Enforcing that `caseSize` in the algorithm produce an equality rather than a subsizing improves this to `tl: (A: Type) -> (l: List^{s8+1} A) -> List^{s8+1} A`, but ultimately still not good enough. We want to be able to say that the domain has the largest size possible, while the range has the smallest size possible. We then need a constraint that imposes `maximize(s0 - s2)`, where `tl: (A: Type) -> (l: List^{s0} A) -> List^{s2} A`. This should result in the solution `tl: (A: Type) -> (l: List^{s8+2} A) -> List^{s8+1} A`. However, there's no way to state this constraint using our current sizes.

**Comment by fbanados** [comment]
Thanks for your note! I guess then this is future work outside of the current scope of the project?  Would size annotations help? (This is a toy example for a more complicated one I had in mind)

Also, is there any plan to add incorporate size annotations?

**Comment by ionathanch** [comment]
Yes, either augmenting the size expressions we can have or adding size annotations are outside the scope. User-added size annotations would *definitely* help -- there's multiple solutions on how a set of size constraints can be solved, and the more information the user provides to indicate what they want the better the solution will actually be what they need. There's no plan to add size annotations to this implementation, and it's unlikely that I ever will...
